WordPress/wporg-gp-translation-events
# Cache list of active events

This PR caches the results of the query that retrieves events that are active "now", so that we're not hitting the database with an expensive query for all actions on a translation.

We use the [`wp_cache_*` functions](https://developer.wordpress.org/reference/classes/wp_object_cache). In development, we use [SQLite Object Cache](https://wordpress.org/plugins/sqlite-object-cache/) as the caching backend.

We cache events for which start is before `now + d` (where `d` is the cache duration) AND end is after `now` (see picture below). We then do extra filtering in code to extract the actually active events.

We cache events that partially or fully overlap with the "cache" box, and we exclude events that don't overlap:

![IMG_20240205_164008 jpg_compressed](https://github.com/WordPress/wporg-gp-translation-events/assets/550401/bce44152-45a8-4fa4-9e4f-41a09bd13d87)


## Testing instructions

1. Checkout `cache-active-events`
2. `composer dev:start`
3. Go to permalink settings and click the save button
4. Create a couple of translation events which last the whole of 2024
    - I'm using the whole year as a example, feel free to use other periods, just make sure they are active "now".
3. Create a translation event which ended before 2024
6. Create a translation event that will start in 2025
7. Place a `die(var_export($events, true));` in an `else` [here](https://github.com/WordPress/wporg-gp-translation-events/blob/cache-active-events/includes/class-wporg-gp-translation-events-translation-listener.php#L138)
8. Go to glotpress and review a translation
    - Make sure you don't add a translation as that might result in an immediate approve, which would mess up with these tests. Instead just reject or approve a "waiting" translation. If there are no waiting translations you can manually change the status of existing entries in `wp_gp_translations` to `waiting`.
10. The `die()` MUST NOT have been called
6. Go to glotpress and review another translation
7. The `die()` MUST have been called and it output the event ids of only the events that are active now.
8. Edit an event
7. The `die()` MUST NOT have been called

## TODO

- [x] Install a [persistent caching plugin](https://developer.wordpress.org/reference/classes/wp_object_cache/#persistent-cache-plugins) in development environment ([SQLite Object Cache](https://wordpress.org/plugins/sqlite-object-cache/))
- [x] Expire cache after a given duration (`d`)
- [x] Cache events for which start is before `now + d` AND end is after `now` (see picture below). Do extra filtering in code to extract the actually active events.
- [x] Invalidate cache when events are created, deleted, or their start, end or timezone are edited.
- [x] Cache event start and end in addition to the id
- [x] Do extra filtering in code to extract the actually active events.
- [x] Add an `Event` class
- [x] Pass `Event` instances to the cache and have the cache return an array of `Event`

## References
- [WP_Object_Cache](https://developer.wordpress.org/reference/classes/wp_object_cache/)
